Analytics add npm install + ci tracking

IgorMinar opened the pull request.

IgorMinar committed:
build(analytics): add support for tracking npm installs

Since the very first npm install is called while node_modules is empty, we need to ignore it, but we can track
the start timestamp and record the install even once the installation is completed.

IgorMinar committed:
build(analytics): add basic ci job tracking

IgorMinar commented:
@alexeagle what do you think about this approach?

I track all start events as well as all end events, but for end events I also track duration.

alexeagle commented on package.json:
```diff
@@ -23,8 +23,8 @@
   "scripts": {
-    "preinstall": "node tools/npm/check-node-modules --purge",
-    "postinstall": "node tools/npm/copy-npm-shrinkwrap && node tools/chromedriverpatch.js && webdriver-manager update && bower install && gulp pubget.dart && tsd reinstall --overwrite --clean --config modules/angular2/tsd.json && tsd reinstall --overwrite --clean --config tools/tsd.json && tsd reinstall --overwrite --config modules/angular1_router/tsd.json",
+    "preinstall": "node tools/analytics/build-analytics start install npm-install; node tools/analytics/build-analytics start install npm-preinstall; node tools/npm/check-node-modules --purge; node tools/analytics/build-analytics finish install npm-preinstall; node tools/analytics/build-analytics start install npm-install-net",
```
can we move this to a `.sh` file? it's getting out of control

IgorMinar replied:
I'd love to but that would make the scripts platform specific. the current approach is the only way to make this work across all platforms.

alexeagle replied:
I figured as much. Here is another idea:
```
 "scripts": {
    "test": "npm run test1 && npm run test2",
    "test1": "echo test1",
    "test2": "echo test2"
  },
```

I tried that out and it appears to work fine. The scripts here will still be long but at least we could split into a few bits, maybe:
```
"postinstall": "node tools/analytics/build-analytics finish install npm-install-net; node tools/analytics/build-analytics start install npm-postinstall; npm run setup-env; npm run install-typings; node tools/analytics/build-analytics finish install npm-postinstall; node tools/analytics/build-analytics finish install npm-install",
```
It would also help with the nesting of the analytics instrumentation. Parent-level timespans can be in the 'master' command and children timespans go in the sub-commands

alexeagle commented on tools/analytics/analytics.js:
```diff
@@ -81,27 +88,51 @@ function getDartVersion() {
+  installStart: (actionName) => {
+    if (ua) {
+      visitor.
+      event('install', actionName + ' (start)', 'testLabel', null, customParams).
```
clang-format?

IgorMinar replied:
yeah

alexeagle commented on .travis.yml:
```diff
@@ -58,23 +60,33 @@ before_install:
 
 script:
+- node tools/analytics/build-analytics start ci script
 - ./scripts/ci/build_and_test.sh ${MODE}
```
of course most of the time is here, we'd like to drill in further - is that possible already with the instrumentation you added earlier?

IgorMinar replied:
yeah, except for protractor tests that currently don't run via gulp.

of course with this PR we can also add more probes anywhere we'd like to get more insights.

alexeagle commented on .travis.yml:
```diff
@@ -58,23 +60,33 @@ before_install:
+- node tools/analytics/build-analytics finish ci script
 
 after_script:
+- node tools/analytics/build-analytics start ci after_script
```
these timings are observable from the log file (which is how http://buildtimetrend.herokuapp.com/dashboard/angular/angular/index.html shows us stuff) but I see the value in getting them to analytics as well.

It would be nicer if some postprocessing on the log files grabbed these timings and pushed them to analytics, since this script is getting more responsibilities and it's easy to mess up the start/finish pairs while maintaining it. But OTOH we don't run the stuff that parses the log files so its not in our control.

IgorMinar replied:
by going through analytics we can correlate timing with hw, node versions, etc. this is way more powerful if setup properly.

rahulkoutu replied:
Igor: I am new to angularjs. Seeking help from your memory game code.
I want to achieve the same using arrow keys and enter button. 
How should I go about it. I am struggling to get focus and key press events. I can do it without angularjs but defeating the whole purpose.

alexeagle replied:
This is certainly not the right way to contact Igor, it's distracting and unwelcome in a pull request review. Can you try the mailing list?

alexeagle replied:
@IgorMinar my comments are driven partly by wishing we had launched tux... if we do get this to be useful, most Travis users could benefit. Maybe one day.

rahulkoutu replied:
I apologize. Not sure how to contact him.

IgorMinar replied:
@rahulkoutu please try https://groups.google.com/forum/#!forum/angular

alexeagle commented on tools/analytics/analytics.js:
```diff
@@ -4,7 +4,13 @@ let execSync = require('child_process').execSync;
+try {
+  ua = require('universal-analytics');
+} catch(e) {
+  // ignore errors due to invoking analytics before the first npm install
```
if we can't get the ua library, is there any point continuing to execute this script? won't we just skip all the logic since it's guarded with `if (ua) {}` or am I missing something useful we do in this case?

IgorMinar replied:
we keep track of the start timestamp for each action. so even if we can't track the beginning in analytics, we can track completion and note duration. I think it's a good enough trade off.

alexeagle replied:
okay so let me see if I understand: each time an event occurs, we run this script, usually the require command is run repeatedly and returns the same thing, and since we stay in the same process we retain the local state including the map of start times.

IgorMinar replied:
the start timestamp is stored in the file under `tmp/analytics/`. that way we can track duration across many node processes.

alexeagle commented on tools/analytics/build-analytics:
```diff
@@ -0,0 +1,51 @@
+
+var startTimestampFilePath = path.resolve(path.join(__dirname, '..', '..', 'tmp', 'analytics', actionCategory + '-' + actionName));
+var analyticsDirPath = path.dirname(startTimestampFilePath);
+var tmpDirPath = path.dirname(analyticsDirPath);
```
did we already have code somewhere else that writes analytics data in tmpfiles?

IgorMinar replied:
in analytics.js, we write unique client id to `.build-analytics` so that we can keep track of sessions over many processes.

alexeagle replied:
k I assume you've factored it to be as common as possible/practical

IgorMinar replied:
I could have used `fs-extras` to make the code shorter, but then the file couldn't be invoked before `npm install`, so I wrote it without any 3rd party libs.

alexeagle commented:
Can you add me to the analytics account to see what we're gathering already?

alexeagle commented:
Nice, I'm sure we can do something with this data, even though it's hard to make Analytics UI show us what we need to know

IgorMinar committed:
build(analytics): track TRAVIS_JOB_NUMBER as a custom dimension

IgorMinar committed:
chore(vscode): add jsconfig.json and remove .settings dir

VSCode recognizes jsconfig.json for es6 code in the same way as it recognizes tsconfig.json for ts code.

mary-poppins commented:
Merging PR #4955 on behalf of @IgorMinar to branch [presubmit-IgorMinar-pr-4955](https://github.com/angular/angular/tree/presubmit-IgorMinar-pr-4955).
